src/engine.rs

```diff
@@ -164,12 +174,35 @@ where
                             len: 0,
                         })
                     };
+                    // If we found that there is no spare space for the next LogBatch in the current
+                    // active file, we will mark the `force_rotate` with `true` to notify the leader
+                    // do `rotate` immediately.
+                    if let Err(Error::Other(e)) = res {
+                        warn!(
+                            "Cannot append, err: {}, try to re-append this log_batch into next log",
+                            e
+                        );
+                        force_rotate = true;
+                        writer.set_output(Err(Error::Other(box_err!(
+                            "Failed to append logbatch, try to dump it to other dir"
+                        ))));
+                        continue;
+                    }
                     writer.set_output(res);
                 }
                 debug_assert!(
                     file_context.id == self.pipe_log.fetch_active_file(LogQueue::Append).id
                 );
                 perf_context!(log_write_duration).observe_since(now);
+                if force_rotate {
+                    if let Err(e) = self.pipe_log.rotate(LogQueue::Append) {
+                        panic!(
+                            "Cannot rotate {:?} queue due to IO error: {}",
+                            LogQueue::Append,
+                            e
+                        );
+                    }
+                }
                 if let Err(e) = self.pipe_log.maybe_sync(LogQueue::Append, sync) {
```

Comment:
It's still incorrect. The NOSPC error would happen here. You need to check its error as well.